blockmanager.go

```diff
@@ -3043,4 +3085,188 @@ func (l *lightHeaderCtx) RelativeAncestorCtx(
 	return newLightHeaderCtx(
 		ancestorHeight, ancestor, l.store, l.headerList,
 	)
+
+}
+
+func (b *blockManager) sideLoadHeaders() {
+
+	// If sideloading is not enabled on this chain
+	// return quickly.
+	if !b.cfg.sideLoad.Enabled {
+		return
+	}
+
+	// If headers contained in the side load source are for a different chain network return
+	// immediately.
+	if b.sideLoadReader.HeadersChain() != b.cfg.ChainParams.Net {
+
+		log.Error("headers from side load file are of network %v "+
+			"and so incompatible with neutrino's current bitcoin network -- skipping side loading", b.sideLoadReader.HeadersChain())
+
+		return
+	}
+	log.Infof("Side loading headers from %v to %v", b.headerTip, b.sideLoadReader.EndHeight())
+
+	// Set headerTip to enable reader supply header, node needs
+	err := b.sideLoadReader.SetHeight(int64(b.headerTip))
+
+	if err != nil {
+		log.Errorf("error while setting height for sideload--- skipping sideloading: "+
+			"%v", err)
+
+		return
+	}
+	headerBatch := make([]headerfs.BlockHeader, 0)
+	for {
+		//Request header
+		header, headerErr := b.sideLoadReader.NextHeader()
+		// If any error occurs while fetching headers that does not indicate
+		// an end of file, return immediately.
+		if headerErr != nil && headerErr != io.EOF {
+
+			log.Errorf("error while fetching headers -- skipping sideloading %v", err)
+
+			return
+		}
+
+		var (
+			node     *headerlist.Node
+			prevNode *headerlist.Node
+		)
+
+		// Update node height if header is not nil
+		if header != nil {
+
+			// Ensure there is a previous header to compare against.
+			prevNodeEl := b.headerList.Back()
+			if prevNodeEl == nil {
+				log.Warnf("side load - Header list does not contain a previous" +
+					"element as expected -- exiting side load")
+
+				return
+			}
+
+			node = &headerlist.Node{Header: *header}
+			prevNode = prevNodeEl
+			node.Height = prevNode.Height + 1
+		}
+
+		if b.cfg.sideLoad.Verify && headerErr == nil {
+			err, passVerification := b.verifyBlockHeader(header, *prevNode)
+			if err != nil || !passVerification {
+				log.Debugf("Side Load- Did not pass verification at height %v"+
+					"-- rolling back to last verified checkpoint and skipping sideload", node.Height)
+
+				prevCheckpoint := b.findPreviousHeaderCheckpoint(
```

Comment:
why are we rolling back to the old checkpoint?

Reply:
Headers received up to a known checkpoint are verified so we can trust those, as opposed to headers received after them.

Reply:
mmm, I would assume that the latest safe block is the one we were before starting loading from the "trusted" source.

Imagine that I download a mobile wallet and synced to the current tip (807675) . I turn off the app, and open it  in one month. I want to download only the new blocks until then. If something goes wrong, I should go back to that block 87675 instead of 295000

Reply:
In this scenario is 295000 a checkpoint? If so, then that is the latest safe block and not the block height before sideloading ever occurred. We can trust blocks before a checkpoint.